docs/design/remerge/rfc.md

```diff
@@ -0,0 +1,938 @@
+They're intended to do as little damage as possible
+
+The input to this is the schema, the local, and incoming remote records.
+
+1. A delta is computed between the local and incoming remote records, in
+   a similar manner to the three-way-merge case. This is known as "the delta".
+
+2. A merged record ("the merged record") is created which starts with all values
+   from the local record not present in "the delta"
+
+    - It doesn't matter if you take from local or remote here, since
+      these are the fields that we just determined were equal.
+
+3. For each composite field containing one or more non-deprecated subfields
+   present in "the delta": the merge is performed roughly the same as the 3WM
+   case.
+
+    - If the composite root has `prefer_remote`, then prefer
+      remote for all members of the composite.
+
+    - If the composite root has `take_newest`, then copy the fields
+      in the composite from whichever has been modified more recently (as
+      determined by modification timestamps) between the local and incoming
+      remote records into "the merged record".
+
+        - Note that we're copying from the records, not from the deltas.
+
+    - If the composite root has `take_min` or `take_max` as it's merge strategy,
+      then compare the values of the composite root in the local and remote
+      records numerically, and copy all fields in the composite from the winner
+      to "the merged record".
+
+4. For remaining (non-composite) fields present in "the delta",
+   "incoming delta", Store the result of the following in "the merged record":
+
+    - For fields with the `take_newest`, `prefer_remote`, `duplicate`,
+      `take_min`, `take_max`, `take_sum`, `prefer_false`, `prefer_true`, and
+      `own_guid` strategy, follow the description listed in the schema format
+      document under "Merge Strategies", noting that this is the two way merge
+      case.
+
+    - For `untyped_map`: The maps are merged directly, breaking ties in favor of
+      the more recently modified.
+        - if `prefer_deletions` is true, any field represented by a tombstone in
+          either side is a tombstone in the output
+
+    - For `record_set`:
+        - The result is the set union of the two, with deletions preferred if
+          `prefer_deletion` is true
+
+5. The "merged delta" is applied to the mirror record to produce the new mirror
+   record which will be uploaded to the server.
+
+    - This record will have a vector clock that is a descendent of the local,
+      mirror, and incoming remote clocks.
+
+## New metadata records:
+
+Some additional per-collection meta-information is required to make remerge
+work.
+
+They are stored at `meta-$collection_name/blah`. This doesn't allow for
+transactional updates at the same time as the records, but in practice so long
+as the schema is always uploaded prior to uploading the records, this should be
+fine.
```

Comment:
The first design of this did that, actually, but it seemed like the legacy collection use case was fairly important, and that we wouldn't be able to update all legacy collections to handle this.

I still like that approach better, TBH, since it's self-contained and can update transactionally.